applications/TrilinosApplication/trilinos_space.h

```diff
@@ -278,41 +294,144 @@ class TrilinosSpace
+     * @param rC The result of the multiplication
+     */
+    static void Mult(
+        const MatrixType& rA,
+        const MatrixType& rB,
+        MatrixType& rC
+        )
+    {
+        const bool transpose_flag = false;
+        EpetraExt::MatrixMatrix::Multiply(rA, transpose_flag, rB, transpose_flag, rC);
+    }
+
     /**
      * @brief Returns the transpose multiplication of a matrix by a vector
      * @details y = AT*x
      * @param rA The matrix considered
      * @param rX The vector considered
      * @param rY The result of the multiplication
      */
-    static void TransposeMult(MatrixType& rA, VectorType& rX, VectorType& rY)
+    static void TransposeMult(
+        const MatrixType& rA,
+        const VectorType& rX,
+        VectorType& rY
+        )
     {
-        //y = A*x
-        bool transpose_flag = true;
+        const bool transpose_flag = true;
         rA.Multiply(transpose_flag, rX, rY);
-    } // rY = rAT * rX
+    }
+
+    /**
+     * @brief Returns the transpose multiplication matrix-matrix
+     * @details C = A*B
+     * @param rA The first matrix considered
+     * @param rB The second matrix considered
+     * @param rC The result of the multiplication
+     * @param TransposeFlag Flags to transpose the matrices
+     */
+    static void TransposeMult(
+        const MatrixType& rA,
+        const MatrixType& rB,
+        MatrixType& rC,
+        const std::vector<bool> TransposeFlag = {false, false}
+        )
+    {
+        KRATOS_ERROR_IF_NOT(TransposeFlag.size() > 1) << "Size of flags must be at least 2" << std::endl;
+        EpetraExt::MatrixMatrix::Multiply(rA, TransposeFlag[0], rB, TransposeFlag[1], rC);
+    }
+
+    /**
+     * @brief Calculates the product operation B'DB
+     * @param rA The resulting matrix
+     * @param rD The "center" matrix
+     * @param rB The matrices to be transposed
+     */
+    static void BtDBProductOperation(
+        MatrixType& rA,
+        const MatrixType& rD,
+        const MatrixType& rB,
+        Epetra_MpiComm& rComm
+        )
+    {
+        // Create a map
+        const int size = Size2(rB);
+        Epetra_Map Map(size, 0, rComm);
+
+        // Create an Epetra_Matrix
+        std::vector<int> NumNz;
+        MatrixType aux(::Copy, Map, NumNz.data());
+
+        TransposeMult(rB, rD, aux, {true, false});
+        Mult(aux, rB, rA);
+    }
+
+    /**
+     * @brief Calculates the product operation BDB'
+     * @param rA The resulting matrix
+     * @param rD The "center" matrix
+     * @param rB The matrices to be transposed
+     */
+    static void BDBtProductOperation(
+        MatrixType& rA,
+        const MatrixType& rD,
+        const MatrixType& rB,
+        Epetra_MpiComm& rComm
+        )
+    {
+        // Create a map
+        const int size = Size1(rB);
+        Epetra_Map Map(size, 0, rComm);
+
+        // Create an Epetra_Matrix
+        std::vector<int> NumNz;
+        MatrixType aux(::Copy, Map, NumNz.data());
```

Comment:
It is not possible do without the Copy?

Iirc Aditya said that this causes memory spikes

Reply:
There are only two options, Copy or View, I will check if it works. In any case this is **an empty matrix**, the memory should be not spiked. If you see old implementations it uses Copy too

Reply:
Done